Setup Sphinx auto documentation, some PEP8 fixes, added docstrings

prasi372 opened the pull request:
## Status

- [ ] Ready
- [x] Draft
- [ ] Hold

## Description

The Sphinx autodoc system is in, along with some PEP-8 fixes and setting docstrings in the right format. Please note that a build scheme to compile and deploy the API docs to a webpage will need to be setup in due course.

## Types of changes

What types of changes does your code introduce to FEDn?

- [ ] Hotfix (fixing a critical bug in production)
- [ ] Bugfix
- [ ] New feature
- [x] Documentation update

## Checklist

_If you're unsure about any of the items below, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] This pull request is against **develop** branch (not applicable for hotfixes)
- [ ] I have included a link to the issue on GitHub or JIRA (if any)
- [ ] I have included migration files (if there are changes to the model classes)
- [ ] I have read the [CONTRIBUTING](https://github.com/scaleoutsystems/fedn/blob/master/CONTRIBUTING.md) doc
- [ ] I have included tests to complement my changes
- [x] I have updated the related documentation (if necessary) 
- [x] I have added a reviewer for this pull request
- [x] I have added myself as an author for this pull request

## Further comments

Anything else you think we should know before merging your code!

prasi372 commented:
In order to build and view the docs, please cd to the apidocs directory, and issue the command 'make html'. Then browse to the build/html directory to open index.html in a browser.

prasi372 commented:
First draft of a unified documentation system.

prasi372 commented:
you may need to:

pip install sphinx_rtd_theme

before doing make html

carmat88 reviewed (changes requested):
Hej @prasi372 First of all, many thanks for this quite conspicuous work! Sphinx is a nice tool.
Almost everything seems to work, except for a couple of things.

When I try to run `make html` within the `apidocs` folder, I get a series of warnings as follows:

![image](https://user-images.githubusercontent.com/26251075/135291373-fd94b2c7-7647-4555-abd7-0bc406544101.png)

Do you get them when compiling with the make command? I think that at least two of them should be correct:

- That is [the page for the FAQ](https://github.com/prasi372/fedn/blob/develop/apidocs/source/faq.md) is still a `.md` file instead for a `.rst`, that is why the make command complain about

`/home/carmat/gitrepo/fedn/apidocs/source/index.rst:11: WARNING: toctree contains reference to nonexisting document 'faq'`

- The other one is related the `module.rst` [file](https://github.com/prasi372/fedn/blob/develop/apidocs/source/modules.rst). It looks like we are never really using it, so my question is: shouldn't we remove it then? Or is it auto generated?

`checking consistency... /home/carmat/gitrepo/fedn/apidocs/source/modules.rst: WARNING: document isn't included in any toctree`

Thank you for looking into the above minor points!

carmat88 commented:
Just for last comment, which it might have to do with just tastes: while I find `sphinx_rtd_theme` easy to recognize within the context of open-source projects, for big screens is a tiny bit unpleasant in the way content is distributed. Look at my entire fullscreen screenshot:

![image](https://user-images.githubusercontent.com/26251075/135295310-642a8fcc-7264-46c6-9667-c7874a53a1ad.png)

It is a responsive design, but I doubt that many would often read such docs on a tablet or mobile phone. So I wonder if we could overwrite the css so to center-align the content. As an alternative here are some themes that looks quite pleasant to read:

- https://sphinx-themes.org/sample-sites/sphinx-press-theme/
- https://sphinx-themes.org/sample-sites/karma-sphinx-theme/
- https://sphinx-themes.org/sample-sites/sphinx-material/
- https://sphinx-themes.org/sample-sites/furo/
- https://sphinx-themes.org/sample-sites/sphinx-book-theme/

My three fav alternatives theme are: sphinx-material, karma-sphinx-theme and sphinx-book-theme. What do you think? Thanks!

carmat88 commented:
Hey @prasi372 ! As per your explanation via discord, the above warnings were related to fedn cli not being installed locally in my system (using a venv, that's why). By just running `pip install fedn` it got fixed:

![image](https://user-images.githubusercontent.com/26251075/135305445-64d1b2ab-e4d3-4e88-a02a-f3449fe4516c.png)

**Note:** I am not getting the warning about the faq document because I have renamed it from `.md` to `.rst` and so it works.

prasi372 commented:
Hi,

Yes I had chosen the sphinx_rtd_theme since it renders fine on phones, tablets and also desktops. The width can be adjusted, and I included a parameter to control that. But your suggested themes look nicer; I liked furo and sphinx_book_theme. So I have switched to the sphinx_book_theme since it was the closest to the feel of the old docs.

Yup, module.rst should not be fiddled with, it allows us to search for modules and associated docs.

I have also now removed the old docs.

In order to test the current PR, you will need to pip install sphinx-book-theme first, and then in the docs folder, make html. Then in the build folder you will be able to see the html files.

carmat88 reviewed (approved):
Looks good!

prasi372 commented on apidocs/source/index.rst:
```diff
@@ -0,0 +1,20 @@
+   sphinx-quickstart on Wed Sep 22 13:30:29 2021.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
```
Yes, that would be good! Nice catch!

carmat88 commented:
> Hi,
> 
> Yes I had chosen the sphinx_rtd_theme since it renders fine on phones, tablets and also desktops. The width can be adjusted, and I included a parameter to control that. But your suggested themes look nicer; I liked furo and sphinx_book_theme. So I have switched to the sphinx_book_theme since it was the closest to the feel of the old docs.
> 
> Yup, module.rst should not be fiddled with, it allows us to search for modules and associated docs.
> 
> I have also now removed the old docs.
> 
> In order to test the current PR, you will need to pip install sphinx-book-theme first, and then in the docs folder, make html. Then in the build folder you will be able to see the html files.

Hey @prasi372 ! The new theme looks nicer, but I think that [this theme](https://sphinx-themes.org/sample-sites/sphinx-press-theme/)  is the closest we have to the previous github doc page: https://scaleoutsystems.github.io/fedn/#/ . Here is a screenshot of how it looks:

![image](https://user-images.githubusercontent.com/26251075/135420643-434206ea-1243-4180-970a-b101daa28408.png)

What do you think? I think that the book theme has confusing double "contents" nav column on both left and right side of the central page.

ahellander commented:
Three is no need to make it close to the old, pick the ones you like most.

Best regards,
Andreas Hellander
Associate Professor, Scientific Computing



30 sep. 2021 kl. 10:55 skrev TechCarmat ***@***.******@***.***>>:



Hi,

Yes I had chosen the sphinx_rtd_theme since it renders fine on phones, tablets and also desktops. The width can be adjusted, and I included a parameter to control that. But your suggested themes look nicer; I liked furo and sphinx_book_theme. So I have switched to the sphinx_book_theme since it was the closest to the feel of the old docs.

Yup, module.rst should not be fiddled with, it allows us to search for modules and associated docs.

I have also now removed the old docs.

In order to test the current PR, you will need to pip install sphinx-book-theme first, and then in the docs folder, make html. Then in the build folder you will be able to see the html files.

Hey @prasi372<https://github.com/prasi372> ! The new theme looks nicer, but I think that this theme<https://sphinx-themes.org/sample-sites/sphinx-press-theme/> is the closest we have to the previous github doc page: https://scaleoutsystems.github.io/fedn/#/ . Here is a screenshot of how it looks:

[image]<https://user-images.githubusercontent.com/26251075/135420643-434206ea-1243-4180-970a-b101daa28408.png>

What do you think? I think that the book theme has confusing double "contents" nav column on both left and right side of the central page.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/scaleoutsystems/fedn/pull/281#issuecomment-931075638>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASBHZSASPS6NKAKR5CFRXLUEQQZNANCNFSM5ERD35VQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.









När du har kontakt med oss på Uppsala universitet med e-post så innebär det att vi behandlar dina personuppgifter. För att läsa mer om hur vi gör det kan du läsa här: http://www.uu.se/om-uu/dataskydd-personuppgifter/

E-mailing Uppsala University means that we will process your personal data. For more information on how this is performed, please read here: http://www.uu.se/en/about-uu/data-protection-policy

prasi372 commented:
I fixed the <no title> thing, but I leave the look and feel to you @carmat88  :) Please feel free to experiment and choose a theme that looks right. Press also looks fine, but didn't work with the title fix.

carmat88 commented:
> I fixed the no title thing, but I leave the look and feel to you @carmat88 :) Please feel free to experiment and choose a theme that looks right. Press also looks fine, but didn't work with the title fix.

Sure, many thanks for all the changes and for setting up the Sphinx structure. I'll proceed to merge and then will take it from our develop branch for the some other look and feel changes.
